daemon/cmd/cni.go

```diff
@@ -0,0 +1,354 @@
+package cmd
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/option"
+
+	"github.com/tidwall/sjson"
+)
+
+// Some functions around managing the CNI configuration file
+//
+// By default, we write the CNI configuration file to disk after the
+// daemon has successfully started up.
+
+// cniConfigs are the default configurations, per chaining mode
+var cniConfigs map[string]string = map[string]string{
+	// the default
+	"none": `
+{
+  "cniVersion": "0.3.1",
+  "name": "cilium",
+  "plugins": [
+    {
+       "type": "cilium-cni",
+       "enable-debug": {{.Debug | js }},
+       "log-file": "{{.LogFile | js }}"
+    }
+  ]
+}`,
+
+	"flannel": `
+{
+  "cniVersion": "0.3.1",
+  "name": "flannel",
+  "plugins": [
+    {
+      "type": "flannel",
+      "delegate": {
+         "hairpinMode": true,
+         "isDefaultGateway": true
+      }
+    },
+    {
+      "type": "portmap",
+      "capabilities": {
+        "portMappings": true
+      }
+    },
+    {
+       "type": "cilium-cni",
+       "enable-debug": {{.Debug | js }},
+       "log-file": "{{.LogFile | js }}"
+    }
+  ]
+}
+`,
+	"portmap": `
+{
+  "cniVersion": "0.3.1",
+  "name": "portmap",
+  "plugins": [
+    {
+       "type": "cilium-cni",
+       "enable-debug": {{.Debug | js }},
+       "log-file": "{{.LogFile | js }}"
+    },
+    {
+      "type": "portmap",
+      "capabilities": {"portMappings": true}
+    }
+  ]
+}
+`,
+}
+
+// The CNI plugin config we inject in to the plugins[] array of existing AWS configs
+var awsCNIEntry = `
+{
+	"type": "cilium-cni", 
+	"enable-debug": {{.Debug | js }}, 
+	"log-file": "{{.LogFile | js }}"
+}
+`
+
+const cniControllerName = "write-cni-file"
+
+func (d *Daemon) startCNIConfWriter(opts *option.DaemonConfig, cleaner *daemonCleanup) {
+	if opts.WriteCNIConfigurationWhenReady == "" {
+		return
+	}
+
+	// We used to disable CNI generation with this environment variable
+	// It's no longer documented, but we need to still support it.
+	if t := os.Getenv("CILIUM_CUSTOM_CNI_CONF"); t == "true" {
+		return
+	}
+
+	d.controllers.UpdateController(cniControllerName,
+		controller.ControllerParams{
+			DoFunc: func(ctx context.Context) error {
+				err := installCNIConfFile(opts)
+				if err != nil {
+					log.Printf("Failed to write CNI config file (will retry): %v", err)
+				}
+				return err
+			},
+			Context:                d.ctx,
+			ErrorRetryBaseDuration: 10 * time.Second,
+		},
+	)
+
+	cleaner.cleanupFuncs.Add(func() {
+		removeCNIConfFile(opts)
+	})
+}
+
+// installCNIConfFile writes the CNI configuration file to disk.
+// This is done once the daemon has started up, to signify to the
+// kubelet that we're ready to handle sandbox creation.
+// There are numerous, conflicting CNI options, exposed in Helm and
+// the cilium-config config map.
```

Comment:
I rewrote this comment to only discuss cilium-config settings, as they're more relevant.